modules/drivers/vertica/src/metabase/driver/vertica.clj

```diff
@@ -138,99 +137,89 @@
-              (hx/cast
-               :integer
-               (hx/floor (hx// (hx/- (datediff :month a b)
-                                     (hx/cast :integer (hsql/call :> (extract :day a) (extract :day b))))
-                               3))))]
-        (hsql/call :case
-                   (hsql/call :<= (cast-timestamp x) (cast-timestamp y))
-                   (positive-diff x y)
-                   :else
-                   (hx/* -1 (positive-diff y x))))
-
-      :month
-      (let [positive-diff (fn [a b]
-                            (hx/-
-                             (datediff :month a b)
-                             (hx/cast
-                              :integer
-                              (hsql/call :> (extract :day a) (extract :day b)))))]
-        (hsql/call :case
-                   (hsql/call :<= (cast-timestamp x) (cast-timestamp y))
-                   (positive-diff x y)
-                   :else
-                   (hx/* -1 (positive-diff y x))))
-
-      :week
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (hx// (datediff :day a b) 7))))]
-        (hsql/call :case
-                   (hsql/call :<= (cast-timestamp x) (cast-timestamp y))
-                   (positive-diff x y)
-                   :else
-                   (hx/* -1 (positive-diff y x))))
-
-      :day
-      (datediff :day x y)
-
-      (:hour :minute :second)
-      (let [positive-diff (fn [a b]
-                            (hx/cast
-                             :integer
-                             (hx/floor
-                              (cond-> (hsql/call :- (extract :epoch b) (extract :epoch a))
-                                (not= unit :second)
-                                (hx// (case unit :hour 3600 :minute 60))))))]
-        (hsql/call :case
-                   (hsql/call :<= (cast-timestamp x) (cast-timestamp y))
-                   (positive-diff x y)
-                   :else
-                   (hx/* -1 (positive-diff y x)))))))
+(defmethod sql.qp/datetime-diff [:vertica :year]
+  [_driver _unit x y]
+  (let [positive-diff (fn [a b] ; precondition: a <= b
+                        (hx/-
+                         (datediff :year a b)
+                         (hx/cast
+                          :integer
+                          (hsql/call
+                           :or
+                           (hsql/call :> (extract :month a) (extract :month b))
+                           (hsql/call
+                            :and
+                            (hsql/call := (extract :month a) (extract :month b))
+                            (hsql/call :> (extract :day a) (extract :day b)))))))]
+    (hsql/call :case
+               (hsql/call :<= (cast-timestamp x) (cast-timestamp y))
+               (positive-diff x y)
+               :else
+               (hx/* -1 (positive-diff y x)))))
```

Comment:
I know this should have comments explaining this complex expression but that's not the focus of this PR.

Reply:
I would have been ok with those in this PR. Either way tho